enhancements/api-review/azure_user_defined_tags.md

```diff
@@ -0,0 +1,301 @@
+    user defined tags to 5 and 10 for Openshift's internal use, for all the resources 
+    created by Openshift.
+
+All in-cluster operators that create Azure resources (Cluster Infrastructure ,Storage ,Node ,NetworkEdge , Internal Registry ,CCO) will apply these tags during resource creation.
+
+The userTags field is intended to be set at install time and is considered immutable. 
+Components that respect this field must only ever add tags that they retrieve from this 
+field to cloud resources, they must never remove tags from the existing underlying cloud 
+resource even if the tags are removed from this field(despite it being immutable).
+
+If the userTags field is changed post-install, there is no guarantee about how an 
+in-cluster operator will respond to the change. Some operators may reconcile the 
+change and change tags on the Azure resource. Some operators may ignore the change. 
+However, if tags are removed from userTags, the tag will not be removed from the 
+Azure resource.
+
+### Workflow Description
+
+- An Openshift administrator requests to add required tags to all Azure resources 
+  created by Openshift by adding it in `.platform.azure.userTags`
+- openshift installer validates the tags defined in `.platform.azure.userTags` and 
+  adds these tags to all resources created during installation and also updates 
+  `.status.platformStatus.azure.resourceTags` of the `infrastructure.config.openshift.io`
+- In cluster operators refers `.sttaus.platformSTatus.azure.resourceTags` of the 
+  `infrastructure.config.openshift.io` to add tags to resources created later.
+
+#### Variation [optional]
+
+### API Extensions
+Enhancement requires below modifications to the mentioned CRDs
+- Add `userTags` field to `platform.azure` of the `installconfigs.install.openshift.io`
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: installconfigs.install.openshift.io
+spec:
+  versions:
+  - name: v1
+    schema:
+      openAPIV3Schema:
+        properties:
+          platform:
+            properties:
+              azure:
+                properties:
+                  userTags:
+                    additionalProperties:
+                      type: string
+                    description: UserTags additional keys and values that the installer
+                      will add as tags to all resources that it creates. Resources
+                      created by the cluster itself may not include these tags.
+                  type: object
+```
+
+- Add `resourceTags` field to `platformStatus.status.azure` 
+  of the `infrastructure.config.openshift.io`
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: infrastructures.config.openshift.io
+status:
+  versions:
+  - name: v1
+    schema:
+      openAPIV3Schema:
+        properties:
+          status:
+            properties:
+              platformStatus:
```

Comment:
The AWS equivalent here is in the status, however, that has caused issues as we now have requests for users to be able to update the tags. I think this should go into the spec, though you'll need to explore the implications of that change as it's different from AWS. There is an enhancement somewhere that discusses the AWS change

Reply:
Thank you for the comment!! 
Yeah, but for Azure and GCP ask in RFE-2017 is for tagging of resources created during Day-0 operation for now. So thought to do it, as it was done for AWS, as including it in Spec, will give out the idea it can edited. Should we add it in spec, to avoid issues we had with AWS and document saying updating userTags will have no effect?

Reply:
> we now have requests for users to be able to update the tags

Do we?  The last conversations i had w/ anyone one the subject, we concluded that create-time only was fine, at least for user tags.  Are you referring to user provided tags or something else?

Reply:
It's possible that I'm out of date here, as far as I knew there was an effort earlier this year to support updating tags on AWS. If the conclusion of that was that we aren't going to do it and we are never going to do it, and we are going to apply the same conclusion here, then it seems reasonable to follow the pattern of putting this in the status. 

Do we have any record of that conversation?

Reply:
> Should we add it in spec to avoid issues we had with AWS and document saying updating userTags will have no effect?

We could put it in the spec and use a webhook to make it immutable, this would provide the same guarantees (technically better) as putting it in status but would allow us to change it later, though removing immutability from an API could be considered a breaking change 🤔

Reply:
If CEL is used, we should have tags in spec and status, and validate if the tags and the tag count matches. And status will be updated regardless of whether all operators successfully applied the tags on the resources it manages, as in the current proposal.
But CEL is supported in 1.25 beta version, can we use the beta version?

Reply:
Beta means it's on by default, so we are using CEL from openshift 4.12 onwards